allow list of search patterns in testFiles config option

bahmutov opened the pull request:
- Closes #5401 

### User facing changelog

- [x] allows a list of specs when searching for test files via config argument `testFiles` which follows the principle in `ignoreTestFiles` config option

### Additional details

Using [`glob`](https://github.com/isaacs/node-glob) to select several groups of spec names is not obvious. For example to select both `.md` and `*spec.js` files one would need to write without spaces

```json
{
  "testFiles": "{*.md,*spec.js}"
}
```

With this change, user could write

```json
{
  "testFiles": ["*.md", "*spec.js"]
}
```

which to me looks more natural and easier to remember. Also added debug log messages during `DEBUG=cypress:server:specs` that show how the search happens in case on needs to debug it.

### How has the user experience changed?

<!--
Provide before and after examples of the change.
Screenshots or GIFs are preferred.
-->

### PR Tasks

<!-- 
These tasks must be completed before a PR is merged.
Delete tasks if they are not applicable. 
-->

- [x] Have tests been added/updated?
- [X] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/2184
- [x] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- [x] Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

bahmutov committed:
WIP: log spec search steps

bahmutov committed:
add multiple files via glob test

bahmutov committed:
allow testFiles to be a string or a list of strings to match

bahmutov commented on packages/server/test/unit/specs_spec.coffee:
```diff
@@ -56,23 +57,48 @@ describe "lib/util/specs", ->
           path.join(cfg.projectRoot, "cypress", "integration", "js_spec.js")
-        ])
+        ]
+        specsUtil.find(cfg, specPattern)
```
being a little bit clearer on the arguments

bahmutov commented on packages/server/test/unit/specs_spec.coffee:
```diff
@@ -56,23 +57,48 @@ describe "lib/util/specs", ->
       config.get(FixturesHelper.projectPath("various-file-types"))
       .then (cfg) ->
-        specsUtil.find(cfg, [
+        debug("test config testFiles is %o", cfg.testFiles)
```
adding a few debug statements to the test itself is useful when iterating on the test and understanding what is going on

bahmutov commented on packages/server/lib/util/specs.js:
```diff
@@ -132,17 +142,31 @@ const find = function (config, specPattern) {
-  .tap((files) => {
-    return debug('found %d spec files: %o', files.length, files)
-  })
+  return Promise.mapSeries(testFilesPatterns, findOnePattern).then(_.flatten)
```
find matching spec files one pattern at a time and then merging them

bahmutov commented on cli/schema/cypress.schema.json:
```diff
@@ -45,7 +45,7 @@
       "type": "string",
       "default": "**/*.*",
-      "description": "A String glob pattern of the test files to load"
+      "description": "A String glob pattern of the test files to load. See https://on.cypress.io/configuration#Global"
```
TODO: needs to be updated before merging

bahmutov committed:
Merge branch 'develop' into list-spec-files-5401

andrew-codes committed:
failing config tests: testFiles type change

Change `testFiles` from a string to either a string **or** array of strings; following similar pattern as `ignoreTestFiles`

andrew-codes committed:
testFiles type change passing tests

andrew-codes committed:
update verbiage related to testFiles type change

andrew-codes committed:
enable running integration tests locally

See https://github.com/paulmillr/chokidar/issues/855 for more details.

andrew-codes committed:
integration tests; failing

Seems to fail for similar reason as #4543 and is likely resolved with PR #4849

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cypress-io/cypress?pullRequest=5402) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/cypress-io/cypress?pullRequest=5402) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: bahmutov<br/>:x: andrew-codes<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cypress-io/cypress?pullRequest=5402) it.</sub>

andrew-codes commented on packages/server/test/integration/cypress_spec.coffee:
```diff
@@ -336,6 +336,28 @@ describe "lib/cypress", ->
+      .then =>
+        expect(browsers.open).to.be.calledWithMatch(ELECTRON_BROWSER, {url: "http://localhost:8888/__/#/tests/integration/test1.js"})
+        @expectExitWith(0)
+
```
@bahmutov per the commit message, I'll echo it here, I think these fail for similar reasons as #4896. There is an outstanding PR for this issue. I can review the PR, get it merged when ready, and then update this branch to see if it resolves this issue.

bahmutov replied:
oh my

andrew-codes replied:
I retract my assessment. It looks like we do **not** need the functionality in that PR. Furthermore, supporting the syntax `--config=testFiles=["**/*.js", "**/*.coffee"]` is how other mutli-value config values work; such as ignoreTestFiles. I do not see any documentation or elsewhere in tests indicating that the format, `--config=testFiles=**/*.js,**/*.coffee`, is actually supported.

andrew-codes committed:
correct test verbiage

Distinguish between the two tests; as they are subtly different.

bahmutov commented on cli/types/cypress-npm-api.d.ts:
```diff
@@ -43,7 +43,7 @@ declare module 'cypress' {
     reporter: string,
     /**
-     * A String glob pattern of the test files to load.
+     * A String or Array of string glob pattern of the test files to load.
```
got to update the type here as well
```
testFiles: string | string[]
```

andrew-codes committed:
update TS types for testFiles type change

Updated the comments, but forgot to update the actual type. This commit rectifies that.

andrew-codes committed:
correct CLI syntax in test

`--config` was not being passed the appropriate value. Now matches other test cases; such as that found in [args_spec](https://github.com/cypress-io/cypress/blob/92b91fe514e5ff6286b4d4e26d2df23062bdf869/packages/server/test/unit/args_spec.coffee#L210)

andrew-codes committed:
remove unsupported syntax from tests

From looking at other tests, it does not appear that the syntax `--config=testFiles=glob1,glob2` is current supported. Removing the test for this test case.

bahmutov commented on packages/server/test/integration/cypress_spec.coffee:
```diff
@@ -336,6 +336,20 @@ describe "lib/cypress", ->
+    it "runs project by limiting spec files via config.testFiles as a JSON array of string glob patterns", ->
+      cypress.start(["--run-project=#{@todosPath}", "--config=testFiles=[\"**/test2.coffee\",\"**/test1.js\"]"])
+      .then =>
+        expect(browsers.open).to.be.calledWithMatch(ELECTRON_BROWSER, {url: "http://localhost:8888/__/#/tests/integration/test2.coffee"})
```
nice

andrew-codes replied:
ty

bahmutov commented on packages/server/test/unit/config_spec.coffee:
```diff
@@ -322,9 +322,18 @@ describe "lib/config", ->
 
-        it "fails if not a string", ->
+        it "passes if an array of strings", ->
+          @setup({ignoreTestFiles: ["**/*.coffee", "**/*.jsx"]})
```
you mean `testFiles` here right ;)

andrew-codes replied:
Yup, let me change these two places within `config_spec`

bahmutov commented on packages/server/test/unit/config_spec.coffee:
```diff
@@ -322,9 +322,18 @@ describe "lib/config", ->
+          @expectValidationFails("be a string or an array of string")
+
+        it "fails if not an array of strings", ->
+          @setup({ignoreTestFiles: [5]})
```
you mean `testFiles`

bahmutov reviewed (commented):
see two comments in tests

andrew-codes committed:
test the correct config value

Previously was testing `ignoreTestFiles` instead of `testFies`

Bkucera commented on cli/schema/cypress.schema.json:
```diff
@@ -45,7 +45,7 @@
       "type": "string",
       "default": "**/*.*",
-      "description": "A String glob pattern of the test files to load"
+      "description": "A String or Array of string glob pattern of the test files to load. See https://on.cypress.io/configuration#Global"
```
doesn't sound right as is, maybe something like
```suggestion
      "description": "A String or Array of glob patterns of test files to load. See https://on.cypress.io/configuration#Global"
```

andrew-codes replied:
Agreed and changed.

andrew-codes committed:
provide more accurate test assertion

Bkucera commented on cli/schema/cypress.schema.json:
```diff
@@ -45,7 +45,7 @@
     "testFiles": {
       "type": "string",
```
```suggestion
      "type": [
        "string",
        "array"
      ],
```

andrew-codes replied:
Good catch! I have updated this.

Bkucera commented on packages/server/test/integration/cypress_spec.coffee:
```diff
@@ -758,6 +772,7 @@ describe "lib/cypress", ->
 
           @expectExitWith(0)
 
+
```
```suggestion
```

andrew-codes committed:
correct config's `testFiles` type

Should be a string or an array of strings.

andrew-codes committed:
remove unnecessary newline

Bkucera commented on packages/server/test/integration/cypress_spec.coffee:
```diff
@@ -336,6 +336,28 @@ describe "lib/cypress", ->
+        @expectExitWith(0)
+
+    it "runs project by limiting spec files via config.testFiles as an array of string glob patterns", ->
+      cypress.start(["--run-project=#{@todosPath}", "--config='[\"testFiles=**/test2.coffee\",\"**/test1.js\"]'"])
```
what is this syntax `--config='["testFiles=`? there is a `[` after `--config=`, which I do not see anywhere else in the tests

andrew-codes replied:
I was not sure what the syntax was supposed to be, but found a similar multi-value configuration option, `blacklistHosts` [here](https://github.com/cypress-io/cypress/blob/92b91fe514e5ff6286b4d4e26d2df23062bdf869/packages/server/test/unit/args_spec.coffee#L140-L148), so I followed that pattern. In the link, it appears that the multi-value value of blacklistHosts is simply JSON.stringified as the CLI option.

If this is not correct, we can change it. I may have found or looked in the wrong places to find the syntax.

The output of this discussion will also impact https://github.com/cypress-io/cypress-documentation/pull/2184

andrew-codes replied:
I think the consensus is to follow the same syntax as `blacklistHosts`; in which case this is the correct syntax: `--config=testFiles=["glob1","glob2"]`.

Note that `spec` and `tag` are run with a slightly different syntax, so we may want to reconsider the syntax across all multi-value values in the future.

Bkucera replied:
notice in that example you've linked, the `[` does not included in the args in the cli

andrew-codes replied:
In hindsight, I don't think my example and how it was conveyed was very clear. Unless I am misunderstanding you, I think the CLI args are: something like this:

`--config=pageLoadTimeout=10000,waitForAnimations=false,hosts={"foobar.com":"127.0.0.1"},blacklistHosts=["one.com","www.two.io"]`

My reasoning is outlined, hopefully more clearly, below:

- `blacklistHosts` is an array [here](https://github.com/cypress-io/cypress/blob/92b91fe514e5ff6286b4d4e26d2df23062bdf869/packages/server/test/unit/args_spec.coffee#L128)
- it is turned into JSON [here](https://github.com/cypress-io/cypress/blob/92b91fe514e5ff6286b4d4e26d2df23062bdf869/packages/server/test/unit/args_spec.coffee#L140)
- that JSON array string is used for the blacklistHosts CLI command [here](https://github.com/cypress-io/cypress/blob/92b91fe514e5ff6286b4d4e26d2df23062bdf869/packages/server/test/unit/args_spec.coffee#L148)
- the `join()` is joining all the CLI arguments; one of which is blacklistHosts which is `blacklistHosts=["one.com", "www.two.io"]`

andrew-codes replied:
Per our discussion, I understand what you are saying now. Resolving this comment and updating code accordingly.

andrew-codes committed:
Merge branch 'develop' into list-spec-files-5401

Bkucera commented on cli/schema/cypress.schema.json:
```diff
@@ -43,9 +43,12 @@
+      ],
       "default": "**/*.*",
-      "description": "A String glob pattern of the test files to load"
+      "description": "A String or Array of glob patterns of the test files to load. See https://on.cypress.io/configuration#Global"
```
```suggestion
      "description": "A String or Array of string glob patterns of the test files to load. See https://on.cypress.io/configuration#Global"
```

kuceb reviewed (dismissed):
accept that wording change if you want. but LGTM

Bkucera commented:
make sure to merge in develop and rebuild after we get this circle zip file size figured out

andrew-codes committed:
tweak verbiage to better align with intention

Co-Authored-By: Ben Kucera <14625260+Bkucera@users.noreply.github.com>

andrew-codes committed:
Merge branch 'develop' into list-spec-files-5401
